src/Orleans.Runtime/Placement/SiloRoleBasedPlacementDirector.cs

```diff
@@ -0,0 +1,39 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Internal;
+using Orleans.Runtime.MembershipService;
+
+namespace Orleans.Runtime.Placement
+{
+    internal class SiloRoleBasedPlacementDirector : IPlacementDirector
+    {
+        private readonly MembershipTableManager membershipTableManager;
+        private static readonly SafeRandom random = new SafeRandom();
+
+        public SiloRoleBasedPlacementDirector(MembershipTableManager membershipTableManager)
+        {
+            this.membershipTableManager = membershipTableManager;
+        }
+
+        public virtual Task<SiloAddress> OnAddActivation(
+            PlacementStrategy strategy, PlacementTarget target, IPlacementContext context)
+        {
+            string siloRole = target.GrainIdentity.PrimaryKeyString;
+
+            List<SiloAddress> siloAddressesSameRole = membershipTableManager.MembershipTableSnapshot.Entries
+                .Where(s => s.Value.Status == SiloStatus.Active && s.Value.RoleName == siloRole)
+                .Select(s => s.Key)
+                .Intersect(context.GetCompatibleSilos(target))
+                .ToList();
+
+            if (siloAddressesSameRole == null || siloAddressesSameRole.Count == 0)
+            {
+                throw new OrleansException($"Available: {membershipTableManager.MembershipTableSnapshot.Entries.FirstOrDefault().Value?.RoleName}. Cannot place grain with RoleName {siloRole}. Either Role name is invalid or there are no active silos with type {siloRole} in MembershipTableSnapshot registered yet.");
```

Comment:
This error message seems a bit off - should the first sentence be there? Additionally, if `FirstOrDefault()` can return null, the safe navigation operator (`?.`) should be used instead of `.` immediately after it

Reply:
Yes, I was planning to remove it. Any suggestion on the proper way of figuring out available RoleNames in test env? 
Initially "testhost" was used and it worked locally and for main. Trying to use 
`var roleName = Assembly.GetEntryAssembly()?.GetName().Name;`
but it doesn't seem to be working based on the latest run.